src/common/core/corerel.js

```diff
@@ -125,6 +133,258 @@ define([
+                node.minimalOverlayShardId = reserveOverlayShard(node);
+            }
+        }
+
+        function updateSmallestOverlayShardIndex(node) {
+            var shardId,
+                minimalItemCount = _shardSize + 1;
+
+            for (shardId in node.overlays) {
+                if (node.overlays[shardId].itemCount < minimalItemCount) {
+                    minimalItemCount = node.overlays[shardId].itemCount;
+                    node.minimalOverlayShardId = shardId;
+                }
+            }
+        }
+
+        function ensureOverlayShardMutated(node, shardId) {
+            var overlayNode;
+
+            if (node.overlayMutations[shardId] !== true) {
+                node.overlayMutations[shardId] = true;
+                node.overlays[shardId] = JSON.parse(JSON.stringify(node.overlays[shardId]));
+                overlayNode = self.getChild(node, CONSTANTS.OVERLAYS_PROPERTY);
+                self.setProperty(overlayNode, shardId, null);
+            }
+        }
+
+        function putEntryIntoOverlayShard(node, shardId, source, name, target) {
+
+            if (node.overlays[shardId].itemCount >= _shardSize &&
+                node.overlays[shardId].items.hasOwnProperty(source) === false) {
+                shardId = reserveOverlayShard(node);
+                node.minimalOverlayShardId = shardId;
+            }
+
+            ensureOverlayShardMutated(node, shardId);
+
+            node.overlays[shardId].items[source] = node.overlays[shardId].items[source] || {};
+            node.overlays[shardId].items[source][name] = target;
+            node.overlays[shardId].itemCount += 1;
+
+            if (node.minimalOverlayShardId === shardId && node.overlays[shardId].itemCount >= _shardSize) {
+                updateSmallestOverlayShardIndex(node);
+            }
+        }
+
+        function putEntryIntoShardedOverlays(node, source, name, target) {
+            // At this point we expect that everything was checked and we can simply look for
+            // the proper place of the entry.
+            var shardId;
+
+            for (shardId in node.overlays) {
+                if (node.overlays[shardId].items.hasOwnProperty(source)) {
+                    putEntryIntoOverlayShard(node, shardId, source, name, target);
+                    return;
+                }
+            }
+
+            putEntryIntoOverlayShard(node, node.minimalOverlayShardId, source, name, target);
+        }
+
+        function removeEntryFromShardedOverlays(node, source, name) {
+            var shardId;
+
+            for (shardId in node.overlays) {
+                if (node.overlays[shardId].items[source] &&
+                    typeof node.overlays[shardId].items[source][name] === 'string') {
+
+                    ensureOverlayShardMutated(node, shardId);
+
+                    delete node.overlays[shardId].items[source][name];
+                    node.overlays[shardId].itemCount -= 1;
+                    break;
```

Comment:
The source entry is being removed during overlayPersist.
If the whole shard become empty, it will not be removed right away, but if it remains empty for the next commit as well..